doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,552 @@
+  - If not cached, need to search WAL after checkpoint and Object meta partition to find the 
+  latest meta data
+
+* Flush (WAL --> Data partition)
+  - Flush WAL entries that have been committed. There are two conditions
+  (1. the size of WAL is close to full, 2. a signal to flush).
+  We can mitigate the overhead of frequent flush via batching processing, but it leads to
+  delaying completion.
+
+
+Crash consistency
+------------------
+* Large case
+  1. Crash occurs right after writing Data blocks
+
+    - Data partition --> | Data blocks |
+    - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+
+  2. Crash occurs right after WAL 
+
+    - Data partition --> | Data blocks |
+    - WAL --> | TxBegin A | Log Entry | TxEnd A |
+    - Write procedure is completed, so there is no data loss or inconsistent state
+
+* Small case
+  1. Crash occurs right after writing WAL
+
+    - WAL --> | TxBegin A | Log Entry| TxEnd A |
+    - All data has been written
+
+
+Comparison
+----------
+* Best case (pre-allocation)
+  - Only need writes on both WAL and Data partition without updating object metadata (for the location).
+* Worst case 
+  - At least three writes are required additionally on WAL, object metadata, and data blocks.
+  - If the flush from WAL to the data parition occurs frequently, radix tree onode structure needs to be update
+  in many times. To minimize such overhead, we can make use of batch processing to minimize the update on the tree
+  (the data related to the object has a locality because it will have the same parent node, so updates can be minimized)
+
+* WAL needs to be flushed if the WAL is close to full or a signal to flush.
+  - The premise behind this design is OSD can manage the latest metadata as a single copy. So,
+  appended entries are not to be read
+* Either best of the worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (the proposed design is similar to LSM-tree which has only level-0)
+  
+
+Detailed Design 
+===============
+
+* Onode lookup
+
+  * Radix tree
+    Our design is entirely based on the prefix tree. Ceph already makes use of the characteristic of OID's prefix to split or search
+    the OID (e.g., pool id + hash + oid). So, the prefix tree fits well to store or search the object. Our scheme is designed 
+    to lookup the prefix tree efficiently.
+
+  * Sharded partition
+    A few bits (leftmost bits of the hash) of the OID determine a sharded partition where the object is located.
+    For example, if the number of partitions is configured as four, The entire space of the hash in hobject_t 
+    can be divided into four domains (0x0xxx ~ 0x3xxx, 0x4xxx ~ 0x7xxx, 0x8xxx ~ 0xBxxx and 0xCxxx ~ 0xFxxx).  
+
+  * Ondisk onode ::
+
+            stuct onode {
+              extent_tree block_maps;
+              b+_tree omaps;
+              b+_tree xattrs;
```

Comment:
xattrs here should simply be inline, I really don't think you want to pay additional lookups for it.  Are you aware of how the osd code (crimson as well at the moment) uses xattrs?

Reply:
Xattr intends to store the small data which can embed in the onode or store EC data . Otherwise, omap is typically larger data than xattr. Right? My intention is that additional space is necessary if inline xattrs become full. Anyway, fixed it.

Reply:
The osd uses an xattr to store the object_info_t on every object and a second one to store the snapset on the head object.  These tend to be accessed on every read and write.  It would be unfortunate to have to do additional reads to access them.

Reply:
I suppose using a btree for the xattrs is fine as long as there's an optimization where they are internal to the onode if they don't take up too much space.

Reply:
I think inline xattr is better.
If xattr is large or the number of xattr  entries grows, the xattr can be moved into the b+tree, omap the onode already has.

Reply:
Any other comments?